src/commands/join.cpp

```diff
@@ -6,119 +6,141 @@
 
 #include <iostream>
 #include <map>
 #include <stdexcept>
 #include <string>
+#include <vector>
 
-#include "../includes/Client.hpp"
-#include "../includes/Server.hpp"
-#include "../includes/colors.hpp"
-#include "../includes/numericReplies.hpp"
-#include "../includes/utils.hpp"
+#include "../../includes/Client.hpp"
+#include "../../includes/Server.hpp"
+#include "../../includes/colors.hpp"
+#include "../../includes/numericReplies.hpp"
+#include "../../includes/utils.hpp"
 
 void Server::joinChannel(const std::string &param, int fd) {
-  // Check if the user is attempting to leave all channels (using "0" as param)
+  Client &client = _clients.at(fd);
+
+  if (param.empty() || (param.length() == 1 && param[0] == REG_CHAN)) {
+    send461NeedMoreParams(client, "JOIN");
+    return;
+  }
   if (isLeaveAllChannelsRequest(param)) {
-    // partAllChannels(fd); // Leave all channels function
+    quitChannel(fd);
     return;
   }
-
-  Client &client = _clients.at(fd);
+  stringVector channels;
   std::string::size_type spacePos = param.find(" ");
   std::string channelsPart = param.substr(0, spacePos);
-  std::string keysPart =
-      (spacePos != std::string::npos) ? param.substr(spacePos + 1) : "";
-  stringVector channels;
   splitByCommaAndTrim(channelsPart, &channels);
-  stringVector keys;
-  splitByCommaAndTrim(keysPart, &keys);
-
-  if (param.empty() || (param.length() == 1 && param[0] == REG_CHAN)) {
-    send461NeedMoreParams(client, "JOIN");
-    return;
+#ifdef DEBUG
+  std::cout << "Before split and trim: " << channelsPart << std::endl;
+  std::cout << "Before split and trim: ";
+  for (size_t i = 0; i < channels.size(); ++i)
+    std::cout << channels[i] << "|";
+  std::cout << std::endl;
+#endif
+  if (isChannelsCorrect(channels, client)) {
+    for (size_t i = 0; i < channels.size(); ++i) {
+      std::string channelNameWithoutPrefix = channels[i].substr(1);
+
+      stringVector keys;
+      std::string keysPart =
+          (spacePos != std::string::npos) ? param.substr(spacePos + 1) : "";
+      splitByCommaAndTrim(keysPart, &keys);
+      checkKeysCorrectness(keys);
+      if (!keys.empty()) {
+        send525InvalidKey(client, getChannelByName(channelNameWithoutPrefix));
+        return;
+      }
+      processJoinRequest(fd, &client, channelNameWithoutPrefix, keys, i);
+    }
   }
+}
 
-  for (size_t i = 0; i < channels.size(); ++i) {
-    std::string channelName = channels[i];
-
-    if (channelName.empty() ||
-        (channelName.length() == 1 && channelName[0] == REG_CHAN)) {
+bool Server::isChannelsCorrect(const stringVector &channels,
+                               const Client &client) {
+  stringVector::const_iterator it = channels.begin();
+  stringVector::const_iterator itEnd = channels.end();
+  for (; it != itEnd; ++it) {
+    if (it->empty() || (it->length() == 1 && (*it)[0] == REG_CHAN)) {
       send461NeedMoreParams(client, "JOIN");
-      continue;
+      return (false);
     }
```

Comment:
Need to add channel name length check with `gConfig->getLimit("CHANNELLEN")`

Reply:
ok, corrigé 

```
    if (it->length() > gConfig->getLimit("CHANNELLEN") || it->length() < 2 || (*it)[0] != REG_CHAN) {
      send476BadChanMask(client, it->substr(1));
```